citation_parser.py

```diff
@@ -0,0 +1,130 @@
+import re
+from collections import namedtuple
+from types import SimpleNamespace
+from nltk.tokenize import sent_tokenize
+
+Citation = namedtuple('Citation', ['text', 'prev', 'start', 'end'])
+
+class CitationParser(object):
+
+    def __init__(self):
+        self._ref_regex = '<ref.+?(/>|</ref>)'
+        self._cn_regex = '{{(C|c)itation needed\|.+?}}'
+        self._citation_regex = self._ref_regex + '|' + self._cn_regex
+        self._citations = []
+        self._sentences = []
+
+    def _find_citations(self, wikitext):
+        for ref in re.finditer(self._citation_regex, wikitext):
+            self._citations.append(Citation(text=ref.group(0), prev=wikitext[ref.start(0)-1], 
+                start=ref.start(0), end=ref.start(0)+len(ref.group(0))))
+        self._citations.reverse()
+
+    def _remove_citation_tags(self, wikitext):
+        return re.sub(self._citation_regex, '', wikitext)
+
+    def _add_successive_citation(self):
+        while self._citations:
+            if self._citations[-1].prev == '>':
+                self._sentences[-1].cite.append(self._citations.pop())
+            else:
+                break
+
+    def parse(self, wikitext):
+        """
+        This is a method for parsing wikitext to get individual sentences
+        and the citations they have.
+        
+        The algorithm goes like:
+        1) Find reference tags <ref> and citation needed templates {cn}
+           within the wikitext and store them in a queue in context order.
+        2) Get the plain text by removing <ref> and {cn}.
+        3) Split the plain text into sentences.
+        4) Loop over the sentences to see whether we can find the sentence
+           in the wikitext.
+           -If found, that means the sentence isn't broken by <ref> or
+            {cn}, then we check whether any citations is followed behind
+            the sentence.
+           -If not found, the sentence is broken by <ref> or {cn}, we find
+            the correct start location and end location of the sentence
+            and also check whether any citation is followed behind.
+
+        """
+        self._find_citations(wikitext)
+        plain_text = self._remove_citation_tags(wikitext)
+        for i, sent in enumerate(sent_tokenize(plain_text)):
+            start = wikitext.find(sent)
+            if start >= 0:
+                # We find the complete sentence in the wikitext, just check if 
+                # any citations is followed.
+                if self._citations and self._citations[-1].start == start+len(sent):
+                    # If a citation is attached behind, the start location
+                    # is equal to the end location of the sentence.
+                    self._sentences.append(SimpleNamespace(text=wikitext[start:start+len(sent)], 
```

Comment:
It's a little odd to define a namedtuple for Citation but just use a SimpleNamespace for sentences. Maybe also have a Sentence namedtuple, or just use a regular tuple here? I don't see why this has to be a SimpleNamespace...

Reply:
SimpleNamespace is new to me actually, so I probably misused it. The reason I don't use namedtuple for sentence is it's immutable but I need to change value afterward. Also SimpleNamespace allows me to use dot "." notation on dictionaries which I think it's convenient.

Reply:
Oh, I see, the immutability is a good point. This is fine then - I think I would have used an actual Sentence class in this case but I don't oppose SimpleNamespace as a simplified dict at all.